Replace ClusterRouterGroup/Pool "use-role" with "use-role-set"

slorion opened the pull request:
When using ```ClusterRouterPool```, we had a need to target routees based on more than one role, ie a routee must have all specified roles. In our use case, we want to select them based on a combination of service area, sub-system and optionally a version, e.g. ["Sales Service","Belgium","v2.0"].

To that end, ```ClusterRouterGroup``` and ```ClusterRouterPool``` can now target routees with a set of roles using ```use-role-set``` instead of ```use-role```.

For now, ```use-role``` has been replaced with ```use-role-set``` everywhere, but of course a backward compatibility layer needs to be added so that ```use-role``` will work as before and be equivalent to ```use-role-set = ["the-role"]```. I am waiting input on how this is usually done in Akka.

Some additional tests for routees selection also need to be added.

slorion committed:
Replace use-role with use-role-set

ClusterRouterGroup/Pool can now target routees with a set of roles (a node must have all specified roles)
TODO: backward compatibility with use-role in the public API (to discuss with Akka team)
TODO: additionnal tests for routee selection

lightbend-cla-validator commented:
Hi @slorion,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

raboof commented:
OK TO TEST

raboof commented:
The test failure is (as expected) our binary compatibility check.

This feature might also be related to #23248, though perhaps not directly.

slorion committed:
Add use-role-set configuration validation

slorion commented:
I made a breaking change by replacing ```useRole: Option[String]``` by ```useRoleSet: Set[String]```. I figured it is simpler to first change the API to what it should be and then add back what is required to stay backward compatible.

The public API is impacted by:

- replacing ```use-role``` (String) by ```use-role-set``` (String array) in the configuration
- replacing ```ClusterRouterGroupSettings.useRole``` by ```useRoleSet```, idem for ```ClusterRouterPoolSettings```

Also, compatibility is impacted by:

- replacing ```useRole``` (optional string) by ```useRoleSet``` (repeated string) in the ```ClusterRouterPoolSettings``` protobuf message definition

patriknw commented on akka-cluster-metrics/src/multi-jvm/scala/akka/cluster/metrics/sample/StatsSampleSpec.scala:
```diff
@@ -45,7 +45,7 @@ object StatsSampleSpecConfig extends MultiNodeConfig {
             enabled = on
             allow-local-routees = on
-            use-role = compute
+            use-role-set = ["compute"]
```
I think it can be named `use-roles` and `useRoles: Set[String]`. We must also consider that there must be a Java API with `java.util.Set`.

patriknw replied:
or we could perhaps use varags in some of the API, which can be used from both java and scala

slorion replied:
I first thought of using ```use-roles```, but I think it does not convey well that it is meant to be all roles that need to be matched, in other words a set of roles. That said, ```use-roles``` may be more idiomatic with Akka configuration nomenclature in general.

As for Java API, would varargs replace existing Java ```Set<String>``` stubs or are they in addition to those stubs ?

I agree that backward compatibility can be challenging without polluting the API. I could add back support for ```use-role```, mark it as deprecated in code and in documentation, only mention ```use-role-set``` ?

patriknw replied:
In many similar we have `with` methods for optional things, `withRoles`.
Here is an example that I worked on the other day, incidentilly also adding support for multi-role: https://github.com/akka/akka/pull/23240/files#diff-2c141d595d12e5b7d90b1931f0cbaf64R130

I guess `ClusterRouterGroupSettings` is changing here. That is even more problematic because it is a `case class`.

slorion replied:
Oh, that's interesting, thanks for the pointer. Would you like me to copy the same design ?

About case classes, is what you mention related to this: https://issues.scala-lang.org/browse/SI-9627 ?

patriknw replied:
yes, kind of. We'll help you with coming up with some workaround (there are ugly tricks).

patriknw replied:
You can start with defining the case class as you would like it to be, with additional `apply` in companion and constructor to cover the old API with `Option[Set]`.

slorion replied:
Based on what I mentioned above (https://github.com/akka/akka/pull/23257#discussion_r125048369) and PR #23240, should I also rename ```use-role-set``` to ```use-roles``` ?

patriknw replied:
I'm not sure role-set is more clear than roles, it would also require docs what it means.
I don't have any better suggestion at the moment

patriknw reviewed (commented):
I think this can be useful. The challenge is to introduce it in a backwards compatible way. We have to keep `use-role` and read from that and `user-role-set`.

slorion committed:
Rename use-role-set with use-roles

slorion committed:
Add backward compatibility with use-role

slorion commented:
@patriknw I renamed ```use-role-set``` to ```use-roles``` and added backward compatibility as best as I could. Please let me know if it is ok and what else needs to be done.

johanandren commented on akka-cluster/src/main/scala/akka/cluster/protobuf/ClusterMessageSerializer.scala:
```diff
@@ -166,8 +166,11 @@ class ClusterMessageSerializer(val system: ExtendedActorSystem) extends BaseSeri
 
+    // for backwards compatibility
     settings.useRole.foreach(builder.setUseRole)
+
```
This means that if using rolling upgrades and a new node has ["a", "b"] and sends to an old node, that will only require the "a" role, no way around that but we should probably document that if using rolling upgrades you should move to the latest versions of Akka first, and then start using multiple roles in a separate rolling upgrade, in the migration guide (I think it would go in `migration-guide-2.4.x-2.5.x.md`)

johanandren commented on akka-cluster/src/main/scala/akka/cluster/protobuf/ClusterMessageSerializer.scala:
```diff
@@ -378,11 +381,12 @@ class ClusterMessageSerializer(val system: ExtendedActorSystem) extends BaseSeri
       maxInstancesPerNode = crps.getMaxInstancesPerNode,
       allowLocalRoutees = crps.getAllowLocalRoutees,
-      useRole = if (crps.hasUseRole) Some(crps.getUseRole) else None
+      useRoles = Option(crps.getUseRole).toSet ++ crps.getUseRolesList.asScala
```
I'd rather se an `if (caps.hasUseRole) caps.getUseRolesList.asScala + caps.getUseRole else ...` here than the extra Option + Set allocations.

johanandren commented on akka-cluster/src/main/resources/reference.conf:
```diff
@@ -239,9 +239,12 @@ akka {
+    # Use members with a particular set of roles, or all members if undefined or empty.
+    use-roles = []
+
+    # For backwards compatibility only, replaced by use-roles
```
`# Deprecated, since Akka 2.5.4, replaced by use-roles`

johanandren commented on akka-cluster/src/main/resources/reference.conf:
```diff
@@ -239,9 +239,12 @@ akka {
     # Useful for master-worker scenario where all routees are remote.
     allow-local-routees = on
 
+    # Use members with a particular set of roles, or all members if undefined or empty.
```
I think it wouldn't hurt to be explicit that a node needs to have all the roles here.

johanandren commented on akka-cluster/src/main/scala/akka/cluster/routing/ClusterRouterConfig.scala:
```diff
@@ -28,14 +28,23 @@ import com.typesafe.config.ConfigFactory
+import scala.collection.JavaConverters._
 
 object ClusterRouterGroupSettings {
+  @deprecated("useRole has been replaced with useRoles")
```
`@deprecated("...", since = "2.5.4")` (and same for all the deprecated annotations)

johanandren commented on akka-cluster/src/main/scala/akka/cluster/routing/ClusterRouterConfig.scala:
```diff
@@ -46,33 +55,46 @@ final case class ClusterRouterGroupSettings(
 
 object ClusterRouterPoolSettings {
+  @deprecated("useRole has been replaced with useRoles")
+  def apply(totalInstances: Int, maxInstancesPerNode: Int, allowLocalRoutees: Boolean, useRole: Option[String]): ClusterRouterPoolSettings =
```
`// for binary compatibility`

johanandren commented on akka-cluster/src/main/scala/akka/cluster/routing/ClusterRouterConfig.scala:
```diff
@@ -46,33 +55,46 @@ final case class ClusterRouterGroupSettings(
   routeesPaths:      immutable.Seq[String],
   allowLocalRoutees: Boolean,
-  useRole:           Option[String]) extends ClusterRouterSettingsBase {
+  useRoles:          Set[String]) extends ClusterRouterSettingsBase {
```
The case class generated copy method now has a different signature, didn't MiMa warn about that?

slorion replied:
I am aware of binary incompatibility and @patriknw also mentioned it, but I do not know how to resolve that issue. Would you mind giving me some pointers ?

johanandren replied:
In `akka.cluster.UniqueAddress` we have done something like this, changing the type of a field, that could be a good thing to look at.

slorion replied:
Ok, I will add copy methods with previous signature and see if it now passes the binary compatibility check.

slorion replied:
Adding 

```scala
    @deprecated("Use constructor with useRoles instead", since = "2.5.4")
    def copy(totalInstances: Int = totalInstances, routeesPaths: immutable.Seq[String] = routeesPaths, allowLocalRoutees: Boolean = allowLocalRoutees, useRole: Option[String] = useRole): ClusterRouterGroupSettings =
        ClusterRouterGroupSettings(totalInstances, routeesPaths, allowLocalRoutees, useRole)
```

prevents the use of the default copy generated for the case class. That means that 
```scala
    def withUseRoles(useRoles: Set[String]): ClusterRouterGroupSettings =
        copy(useRoles = useRoles)
``` 
does not compile. Is that what you have in mind ?

johanandren replied:
I think you would have to do that yes. I guess this means you cannot use the copy method in `withUseRoles` but must rather construct a new instance manually. Isn't keeping binary compatibility great? ;)

slorion replied:
Indeed ... I saw this related Scala issue: https://issues.scala-lang.org/browse/SI-9627. Wish that would be available.

johanandren commented on akka-cluster/src/main/scala/akka/cluster/routing/ClusterRouterConfig.scala:
```diff
@@ -46,33 +55,46 @@ final case class ClusterRouterGroupSettings(
     throw new IllegalArgumentException("routeesPaths must be defined")
 
-  routeesPaths.foreach(p ⇒ p match {
+  routeesPaths.foreach({
```
skip the extra parenthesises?

johanandren commented on akka-cluster/src/main/scala/akka/cluster/routing/ClusterRouterConfig.scala:
```diff
@@ -46,33 +55,46 @@ final case class ClusterRouterGroupSettings(
+  def apply(totalInstances: Int, maxInstancesPerNode: Int, allowLocalRoutees: Boolean, useRole: Option[String]): ClusterRouterPoolSettings =
+    ClusterRouterPoolSettings(totalInstances, maxInstancesPerNode, allowLocalRoutees, useRole.toSet)
+
+  def apply(totalInstances: Int, maxInstancesPerNode: Int, allowLocalRoutees: Boolean, useRoles: String*): ClusterRouterPoolSettings =
```
I think a `@varargs` is needed to make a vararg forwarder from Java

johanandren commented on akka-cluster/src/main/scala/akka/cluster/routing/ClusterRouterConfig.scala:
```diff
@@ -157,8 +181,8 @@ final case class ClusterRouterGroup(local: Group, settings: ClusterRouterGroupSe
-    case ClusterRouterGroup(local, _) ⇒
-      copy(local = this.local.withFallback(local).asInstanceOf[Group])
+    case ClusterRouterGroup(otherLocal, _) ⇒
+      copy(local = this.local.withFallback(otherLocal).asInstanceOf[Group])
```
👍

johanandren commented on akka-cluster/src/test/scala/akka/cluster/protobuf/ClusterMessageSerializerSpec.scala:
```diff
@@ -86,7 +86,7 @@ class ClusterMessageSerializerSpec extends AkkaSpec(
           maxInstancesPerNode = 5,
           allowLocalRoutees = true,
-          useRole = Some("Richard, Duke of Gloucester")
+          useRoles = Set("Richard, Duke of Gloucester")
```
Cover no roles and multiple roles as well?

It would also be good to have backwards bin-comp test coverage, deserialising an old message, you can see how to do that in https://github.com/akka/akka/blob/3643f18ded0cf88dfc098d707512b762fd1ff0f7/akka-remote/src/test/scala/akka/remote/serialization/DaemonMsgCreateSerializerSpec.scala#L108

johanandren commented on akka-docs/src/test/java/jdocs/cluster/FactorialFrontend.java:
```diff
@@ -77,12 +79,12 @@ void sendJobs() {
       HeapMetricsSelector.getInstance(), Collections.<String> emptyList()),
         new ClusterRouterGroupSettings(totalInstances, routeesPaths,
-          allowLocalRoutees, useRole)).props(), "factorialBackendRouter2");
+          allowLocalRoutees, useRoles)).props(), "factorialBackendRouter2");
```
Vararg all of these and you should be able to keep the original sample code

slorion replied:
Adding `@varargs` to constructors produce this compilation error: ```A constructor cannot be annotated with a `varargs` annotation.```

johanandren replied:
Ah, didn't know. That's a shame.

johanandren reviewed (commented):
Some feedback added, looks promising

slorion committed:
Integrate recommendations made by @johanandren

slorion commented:
Is there a way to get the result of the Jenkins tests ? It would make it easier for me to see what exactly I need to work on to pass the validation.

jlprat commented:
You can click next to the Jenkins validation, it should bring you to the build result

jlprat commented:
The `Details` link. The last run is https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/9910/

slorion commented:
I was getting a timeout with that link when on the corporate network,
probably because of the firewall. It works at home, so I will take a look
this evening. Thanks!

On Wed, Aug 2, 2017, 18:19 Josep Prat <notifications@github.com> wrote:

> The Details link. The last run is
> https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/9910/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka/pull/23257#issuecomment-319723036>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABiEEBp28eiL4_ApKxG9rp-4SAZ4rk12ks5sUKF6gaJpZM4OJjtm>
> .
>

slorion commented:
The only error I see is ```(akka-cluster/genjavadoc:doc) javadoc returned nonzero exit code``` and 100 warnings related to `ReliableProxy` and `TimerBasedThrottler`. Not sure what to fix ...

johanandren commented:
Might be this genjavadoc bug: https://github.com/typesafehub/genjavadoc/issues/85

Please try making the scaladoc entries where there is a `@deprecated` annotation multiline. You can test if it fixes things locally by running: `sbt -Dakka.genjavadoc.enabled=true javaunidoc:doc`

lightbend-cla-validator commented:
Hi @xdev-developer,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

lightbend-cla-validator commented:
Hi @xdev-developer,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

slorion committed:
Update javadoc to avoid bug with genjavadoc (typesafehub/genjavadoc#85)

johanandren commented:
Weird stuff going on with the CLA validator. I wonder what's up with that.

johanandren commented:
Seems to be a related test failure: https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/9915/testReport/akka.cluster/ClusterDeployerSpec/_It_is_not_a_test_it_is_a_sbt_testing_SuiteSelector_/

`requirement failed: All roles in useRoles must be non-empty`

slorion committed:
Fix mishandling of empty use-role config parameter

slorion commented:
@johanandren Thanks for pointing out the bug with genjavadoc, would have otherwise been hard to figure this one out!

johanandren commented:
MiMa binary comp problems:

```
[error]  * method useRole()scala.Option in trait akka.cluster.routing.ClusterRouterSettingsBase does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.routing.ClusterRouterSettingsBase.useRole")
[error]  * method useRoles()scala.collection.immutable.Set in trait akka.cluster.routing.ClusterRouterSettingsBase is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.cluster.routing.ClusterRouterSettingsBase.useRoles")
```

Since `ClusterRouterSettingsBase` is internal and the two concrete types are final, it should be safe to add these filters.

slorion commented:
@johanandren Ok, I saw this log and was trying to find a way to make to trait compatible, but if you tell me all I need to do is add exclusions, then that's even better!

slorion committed:
Add exclusions to MiMa check

slorion commented:
The new exclusion file (`2.5.4.backwards.excludes`) seems to be ignored. MiMa is checking against Akka 2.5.1, so should I name the exclusion file `2.5.1.backwards.excludes` ?

slorion committed:
Add some more MiMa exclusions

slorion committed:
Add more serialization unit tests, incl. for old wire format

slorion committed:
Update migration guide with a warning on rolling updgrades

slorion commented:
I think this PR is now ready. Please let me know if there's anything else to be done before it can be merged. 

@johanandren @patriknw @jlprat I take the occasion to say that I am very impressed by the friendliness, support and technical expertise that I experienced doing this contribution. It is my first PR on Akka and it was absolutely painless, thanks to you all!

johanandren reviewed (approved):
LGTM, excellent work on this @slorion !

patriknw reviewed (approved):
LGTM, difficult to get the binary compat right, looking good

slorion commented:
Great! Do you have an idea of which release will contain this feature ? My understanding is 2.5.4, but if later, I would like to know as our architecture depends on this feature being present. Thank you !

johanandren commented:
Fixes #23496
